### PR TITLE
Actions.EasyMotion feature added

### PR DESCRIPTION
**Pull request by partiallyordered**
### Description

Added a new feature to select visible windows with key chords.

This is pretty much the first Haskell I've written, so any critique/suggestions/improvements will be very much welcome. Thanks!

### Checklist

  - [*] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [*] I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

  - [*] I updated the `CHANGES.md` file

### CODE REVIEW

**Comment by msk-** [comment]
See #220, turns out @larkery and I have implemented similar features.

**Comment by anthraxx** [comment]
Any news on this? This looks incredibly nice and it would be awesome to see it getting pulled

**Comment by msk-** [comment]
I'm the worst. A few days ago I fixed up the compile warnings. I haven't had a chance to ensure I didn't break anything yet. Aiming to do that tomorrow. Sorry to anyone who's been following this :(.

**Comment by slotThe** [comment]
This is incredible; @msk- are you still interesting in getting this merged?  If yes then I will try to find some time this as thoroughly as I can

**Comment by msk-** [comment]
@slotThe yup, still interested, a review would be great please :+1:

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:229]: Because you're only using `currentW` to get to the current focus below, you can do 
```suggestion
  let currentW = W.peek $ ws
```

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:232]: `sequence . fmap` already has a name: `mapM`

The "newer" way is to use the slightly more general `traverse = sequenceA . fmap`.

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:250]: Some of these would perhaps be more readable in a `where` clause (and with space separating individual functions)

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:256]: This has the same problem that was fixed for X.Prompt and X.A.TreeSelect in #447, i.e. if I use the following config 

``` haskell
import XMonad
import XMonad.Actions.EasyMotion (selectWindow)
import XMonad.Util.EZConfig (additionalKeysP)

import qualified XMonad.StackSet as W

main :: IO ()
main = xmonad $ def{ clickJustFocuses = False }
  `additionalKeysP`
    [("M-n", selectWindow def >>= (`whenJust` windows . W.focusWindow))]
```

and  spawn two terminals, then press `M-n` and click into the non-focused window, xmonad will grab the keyboard and mouse and never release them.  This should be fixable with a similar setup as #447 in your `event` function

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:276]:
```suggestion
      sanitiseKeys = toList . fromList . L.filter (`notElem` [cancelKey conf, xK_BackSpace])
```

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:267]: I wonder why we don't just remove duplicates every time?  Might need to write that down in the docs, but right now I don't see a reason not to

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:285]: Can you describe, in words, what this number is supposed to represent?

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:331]:
```suggestion
  paintAndWrite w f (fi (rect_width r)) (fi (rect_height r)) (fi brW) bgC brC textC bgC [AlignCenter] [concatMap keysymToString c]
```

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:298]: This is not exported and you're only using it in conjunction with `Overlay`; is there any reason why `a` is a type variable here?

**Review by slotThe — CHANGES_REQUESTED** [review]
Awesome!

So the first round is complete (this will probably take multiple iterations, because the module itself is quite complex; sorry about that)

Some general things I've noticed: 

  1. Instead of merging `master` into your feature branch, you should rebase on top of `master` instead.  Otherwise the history will be incredibly hard to read.  See [contributing](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md#rebasing-and-squashing-commits) for more information
  2. You should also mention your module in `XMonad.Doc.Extending`, so users can find it

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs:87]: Haven't tested this, but a scan of `initXMF` looks as though it'll fall back to `-misc-fixed-*-*-*-*-10-*-*-*-*-*-*-*`. Using core fonts is a pretty bad experience, in my testing. For whatever reason, core fonts cause the overlay to be very slow to appear. In any case, if someone still wants to use the module with core fonts, well, great. The fallback font is _tiny_ for the purpose of this module- a better default is in order at least. It looks as though I might be able to test whether XMonad was compiled with XFT using `#ifdef XFT` and use a conditional default. What do you think?

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs:94]: Is this a convention, or does it have some bearing on the generated docs?

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs:77]: Yeah I think that's what we ended up deciding on in #349 as well; as much as I dislike CPP it's probably the correct choice in this situation

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs]: This is the haddock syntax for monospacing a piece of text (see [here](https://www.haskell.org/haddock/doc/html/ch03s08.html)); so it does have bearing on the generated docs. 

You can also put it on it's own line entirely like this: 

``` haskell
-- Customise the overlay by supplying a function to do so. The signature is @
-- 
-- > 'Position' -> 'Rectangle' -> 'X' 'Rectangle'
-- 
-- The parameters are the height in pixels of the selection chord
```

Whatever you prefer, really, but I do think we should follow the convention that code be monospaced

Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: Useful link, thank you

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: Latent shame and self-loathing. Hah. For the record, I appreciate how kindly you asked this, though I would've probably taken "this is not a good variable name" in good faith.

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs:184]: Is `Maybe Int` appropriate? Feels sorta funny, not sure.

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: Sheesh, I've made.. some.. changes.. I don't have a good feel for Haskell aesthetics yet. E.g. `appendChords` is off by itself, it's a little larger than the others I've put in a `where` clause, but not by much. I'll resolve this particular item for now because the code has changed quite a bit since the above comment. More feedback in this vein very welcome, though.

**Comment by partiallyordered** [comment]
@slotThe thanks a lot for the review, really. It's been a reintroduction to Haskell for me. But that's been good, if a little bit of a trial coming from my js day job. I've addressed most of the feedback, but I suspect you were right about another round- there have been a couple of significant changes and I'm expecting I have more to learn (about.. err.. perhaps monad transformers..?).

Regarding commit history, I'm not sure it'll be very useful, even if linear- unless you object I'll probably just squash it all into "added EasyMotion module" or similar.

Thanks again, it's been a good (re-)learning experience so far.

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs:184]: I would say that that's fine.  If this feels too contrived to you, you can probably just define some type that's isomorphic to `Maybe Int` yourself and give it a better name, like `data MaxLength = NoMaximum | Max Int` or something similar.

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs]: I think `appendChords` being a top-level function is not that bad; don't really want `handleSelectWindow` to become even bigger!  Btw, because I was just looking at this, you should be able to write 

``` Haskell
zipWith Overlay overlayWins chords
```

instead of having to explicitly deal with record syntax there.

Regarding the style I'll jot down a few things in the second review; particularly about those 150+ column lines ;)

**Comment by slotThe** [comment]
> @slotThe thanks a lot for the review, really. 

My pleasure!

> I've addressed most of the feedback, but I suspect you were right about another round- there have been a couple of significant changes and I'm expecting I have more to learn (about.. err.. perhaps monad transformers..?).

Every time you've written a function within the `X` monad you were using monad transformers!  Easy peasy :)  I have some real-life stuff to take care of right now, so the review will probably be delayed by a few weeks; though I promise I'll get to it by early April!

> Regarding commit history, I'm not sure it'll be very useful, even if linear- unless you object I'll probably just squash it all into "added EasyMotion module" or similar.

Yeah I think this is the best approach;  it's nice (for me) to see the progression on what has changed, but the changes were so extensive that I'll probably end up going through the whole module again anyways.  And after merging this I think these would not be _that_ helpful for someone reading over the commit logs.

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:243]: I feel like this may be nicer as documentation for the definition of `f'`

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:269]: I would indent this (just for readability) a little more like 

``` haskell
  overlays <- case sKeys c of
    AnyKeys ks -> buildOverlays ks <$> sortedOverlayWindows
     where
      visibleWindows :: [Window]
      [...]
```

Much less wasted horizontal space.  Also, Haskell people are quite fond of aligning things instead of having long lines, i.e. instead of 

``` Haskell
PerScreenKeys m -> fmap concat $ sequence $ M.foldr (:) [] $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
```

you could write

``` haskell
PerScreenKeys m ->
  fmap concat
    $ sequence
    $ M.foldr (:) []
    $ M.mapWithKey (\sid ks -> buildOverlays ks <$> sortedOverlayWindows sid) m
```

This is your module, so do what you feel like fits best; we're way beyond enforcing any kind of style guide around here :)

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [inline XMonad/Actions/EasyMotion.hs:270]: With `{-# LANGUAGE ScopedTypeVariables #-}` you can make this type signature much more readable and write `overlays :: X [Overlay] <- case ...` directly.

**Review by slotThe — APPROVED** [review]
This looks really good now!  One could try to make this looks more like "standard" Haskell but honestly I'm at this point more interested in finally getting this merged in a timely fashion and pushing a quick fix pr afterwards if there's any need for that.  I just noticed a few more things that I feel like you should at least have a chance to see before this is merged, but it's nothing pressing really.  

I'll merge this in a few days if nothing comes up; thank you for the contribution!

**Comment by partiallyordered** [comment]
@slotThe thanks again. I'm super, super busy this week. I'd like to make some of the changes you've suggested but most likely I'll get to them early next week. If you'd prefer to merge before then, go ahead. (I guess you'll squash, so I won't?).

P.S., my comment about learning monad transformers was based on a suspicion that I could be doing something a little more elegant than using `io` for about half of the lines in a couple of the main bits of handling, e.g.
https://github.com/xmonad/xmonad-contrib/blob/d52f49066405fe298ea5f6407ebabebbc1604fa0/XMonad/Actions/EasyMotion.hs#L271-L277
and
https://github.com/xmonad/xmonad-contrib/blob/d52f49066405fe298ea5f6407ebabebbc1604fa0/XMonad/Actions/EasyMotion.hs#L351-L361
Any tips would be very welcome (totally acknowledging you have no obligation!). Even "yeah, read this other module" would be great. Don't worry, I won't go messing around with what's already in the PR now, just want to learn better for future usage.

**Comment by slotThe** [comment]
> @slotThe thanks again. I'm super, super busy this week. I'd like to make some of the changes you've suggested but most likely I'll get to them early next week. If you'd prefer to merge before then, go ahead. 

That's totally fine.  If you want to look at this a bit more then I'll just wait until you're satisfied—just don't feel pressured to do anything if you don't want to or don't have the time!

> (I guess you'll squash, so I won't?).

Sure, sounds like a plan.  

> P.S., my comment about learning monad transformers was based on a suspicion that I could be doing something a little more elegant than using `io` for about half of the lines in a couple of the main bits of handling, e.g.
> 
> https://github.com/xmonad/xmonad-contrib/blob/d52f49066405fe298ea5f6407ebabebbc1604fa0/XMonad/Actions/EasyMotion.hs#L271-L277
> 
> and
> 
> https://github.com/xmonad/xmonad-contrib/blob/d52f49066405fe298ea5f6407ebabebbc1604fa0/XMonad/Actions/EasyMotion.hs#L351-L361
> 
> Any tips would be very welcome (totally acknowledging you have no obligation!). Even "yeah, read this other module" would be great. Don't worry, I won't go messing around with what's already in the PR now, just want to learn better for future usage.

To be honest, lifting `IO` actions is just something that you sometimes have to do when working with monad transformers (thankfully, `IO` is the only thing you have to lift).  One could argue that this should perhaps be separated as much as possible, but that seems rather hard for contrib in particular; `X` effects want to mingle with what X11 does quite often.  

Arguably, all of the functions that the X11 library exports should have been `MonadIO` instead of plain `IO` (perhaps with a `specialise` pragma for `IO` to avoid unecessary runtime dictionary passing) so we wouldn't have to do this kind of lifting at all.

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: I'm not sure if you meant dedented, I suspect this comment has suffered from autoindent (and my subsequent negligence). I've dedented it, let me know if you meant something else.

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: Regarding the first suggestion: I prefer that too- I suspect this is another autoindent casualty, except in this case I probably would've assumed my editor indent was accepted style, rather than just having made the mistake of not noticing it!

Regarding the second: much nicer, thanks.

File: XMonad/Actions/EasyMotion.hs
Comment (partiallyordered) [reply XMonad/Actions/EasyMotion.hs]: Took a while with the compiler patiently repeating at me:
> `overlays` IS `[Overlay]`, NOT `X [Overlay]`
before I finally listened to it. Hah.

I can't remember if it was this type signature or another one, but I definitely had some frustration fighting correct type signature indentation. And in fact I tried this style to see if it would work, but of course it didn't. `ScopedTypeVariables` is _great_ to know about, thank you.

**Comment by partiallyordered** [comment]
Phew, sorry I'm so overdue (in the past couple of weeks, and for.. years..) addressing this. Thanks again for the feedback, and thanks for answering my questions. I'm happy for this to be merged whenever you are.

**Comment by slotThe** [comment]
:heart:

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs]: You're completely right; thanks!

File: XMonad/Actions/EasyMotion.hs
Comment (slotThe) [reply XMonad/Actions/EasyMotion.hs]: Whoops, sorry about that :)

It's a bit unintuitive that `ScopedTypeVariables` should introduce this behaviour I agree, but it definitely is a very useful extension once you know about it

**Pull request merged**
